src/js/common/parser.js

```diff
@@ -0,0 +1,281 @@
+    var requireDefinitions = that.resolveJsonPointer(schemaKey, failurePointer);
+    var requirementIndex = fluid.find(requireDefinitions, function (value, index) {
+        if (value === propertyToMatch) { return index; }
+    });
+
+    // The path to the message is relative to the parent
+    if (requirementIndex !== undefined && requirementIndex !== null) {
+        var parentPointer = gpii.schema.parser.getParentJsonPointer(failurePointer);
+        var errorsPointer = gpii.schema.parser.getChildJsonPointer(parentPointer, "required");
+        return gpii.schema.parser.getChildJsonPointer(errorsPointer, requirementIndex);
+    }
+    else {
+        return undefined;
+    }
+};
+
+/**
+ *
+ * A static function to return the last segment of a given JSON pointer.
+ *
+ * @param jsonPointer {String} - The original JSON pointer.
+ * @returns A {String} representing the last segment of the original pointer.
+ */
+gpii.schema.parser.getLastJsonPointerSegment = function (jsonPointer) {
+    var segments = jsonPointer.split("/");
+    return segments[segments.length - 1];
+};
+
+/**
+ *
+ * A model listener to dereference and cache a dereferenced version of schemas as they are added to `model.schemas`.
+ */
+gpii.schema.parser.updateSchemas = function (that) {
+    var promises = [];
+    fluid.each(
+        that.model.schemas, function (schemaContent, schemaKey) {
+        if (!that.dereferencedSchemas[schemaKey]) {
+            promises.push(that.dereferenceSchema(schemaKey));
+        }
+    });
+
+    if (promises.length > 0) {
+        fluid.promise.sequence(promises).then(
+            function () { that.events.onSchemasUpdated.fire(that); },
+            function (error) { fluid.fail(error.message || error); }
+        );
+    }
+};
+
+fluid.defaults("gpii.schema.parser", {
+    gradeNames:    ["fluid.modelComponent"],
+    requiredRegexp: "/required$",
+    parserOptions: {},
+    uriTemplate:   "%schemaPath/%schemaKey",
+    messages: {
+        required: "You must provide a value for this field."
+    },
+    model: {
+        schemas: {}
+    },
+    events: {
+        onSchemasUpdated: null
+    },
+    members: {
+        parser: null,
+        dereferencedSchemas: {},
+        isReady: false
+    },
+    invokers: {
+        dereferenceSchema: {
+            funcName: "gpii.schema.parser.dereferenceSchema",
+            args: ["{that}", "{that}.options.schemaPath", "{arguments}.0"] // schemaPath, callback
+        },
+        resolveJsonPointer: {
+            funcName: "gpii.schema.parser.resolveJsonPointer",
+            args:     ["{that}", "{arguments}.0", "{arguments}.1"] // schemaKey, jsonPointer
+        },
+        evolveError: {
+            funcName: "gpii.schema.parser.evolveError",
+            args:     ["{that}", "{arguments}.0", "{arguments}.1"] // schemaKey, jsonPointer
+        }
+    },
+    listeners: {
+        "onSchemasUpdated.indicateReadiness": {
```

Comment:
I think that this event's semantics are a bit buggy. It can't necessarily be advising the user that the component is *indeed* ready to use since several requests for schema updates may have come in since the last time the flag was set. I think the confusion arises from the use of model-like semantics for the "schemas" which imply that they could be updated at any time after the component is constructed. You could either back this out and make a simpler component that accepts all its schemas at construction time, or you else you will have to go forwards and go the whole hog with a fully dynamic component resolving all the races commented above. If you want to do this, the component will have to maintain a "queue" of currently outstanding requests for schemas which is updated by the modelListener below, "isReady" will have to become a model flag that can become false if the queue is nonempty, etc. and then only become set again once the queue drains successfully.